rust/agama-network/src/model.rs

```diff
@@ -0,0 +1,439 @@
+            3 => Ok(DeviceType::Unknown),
+            _ => Err(NetworkStateError::InvalidDeviceType(value)),
+        }
+    }
+}
+
+/// Represents an available network connection
+#[derive(Debug, PartialEq, Clone)]
+pub enum Connection {
+    Ethernet(EthernetConnection),
+    Wireless(WirelessConnection),
+    Loopback(LoopbackConnection),
+}
+
+impl Connection {
+    pub fn new(id: String, device_type: DeviceType) -> Self {
+        let base = BaseConnection {
+            id: id.to_string(),
+            ..Default::default()
+        };
+        match device_type {
+            DeviceType::Wireless => Connection::Wireless(WirelessConnection {
+                base,
+                ..Default::default()
+            }),
+            _ => Connection::Ethernet(EthernetConnection { base }),
+        }
+    }
+
+    pub fn base(&self) -> &BaseConnection {
+        match &self {
+            Connection::Ethernet(conn) => &conn.base,
+            Connection::Wireless(conn) => &conn.base,
+            Connection::Loopback(conn) => &conn.base,
+        }
+    }
+
+    pub fn base_mut(&mut self) -> &mut BaseConnection {
+        match self {
+            Connection::Ethernet(conn) => &mut conn.base,
+            Connection::Wireless(conn) => &mut conn.base,
+            Connection::Loopback(conn) => &mut conn.base,
+        }
+    }
+
+    pub fn id(&self) -> &str {
+        self.base().id.as_str()
+    }
+
+    pub fn uuid(&self) -> Uuid {
+        self.base().uuid
+    }
+
+    pub fn ipv4(&self) -> &Ipv4Config {
+        &self.base().ipv4
+    }
+
+    pub fn ipv4_mut(&mut self) -> &mut Ipv4Config {
+        &mut self.base_mut().ipv4
+    }
+
+    pub fn remove(&mut self) {
+        self.base_mut().status = Status::Missing;
+    }
+
+    pub fn is_missing(&self) -> bool {
+        self.base().status == Status::Missing
+    }
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct BaseConnection {
+    pub id: String,
+    pub uuid: Uuid,
+    pub ipv4: Ipv4Config,
+    pub status: Status,
+}
+
+impl PartialEq for BaseConnection {
+    fn eq(&self, other: &Self) -> bool {
+        self.id == other.id && self.uuid == other.uuid && self.ipv4 == other.ipv4
+    }
+}
+
+#[derive(Debug, Default, Clone, Copy, PartialEq)]
+pub enum Status {
+    #[default]
+    Present,
+    Missing,
+}
+
+#[derive(Debug, Default, PartialEq, Clone)]
+pub struct Ipv4Config {
+    pub method: IpMethod,
+    pub addresses: Vec<(Ipv4Addr, u32)>,
+    pub nameservers: Vec<Ipv4Addr>,
+    pub gateway: Option<Ipv4Addr>,
+}
+
+#[derive(Debug, Default, Copy, Clone, PartialEq)]
+pub enum IpMethod {
+    #[default]
+    Auto = 0,
+    Manual = 1,
+    Unknown = 2,
+}
+impl fmt::Display for IpMethod {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let name = match &self {
+            IpMethod::Auto => "auto",
+            IpMethod::Manual => "manual",
+            IpMethod::Unknown => "auto",
+        };
+        write!(f, "{}", name)
```

Comment:
why not `f.write_str(name)` ?

Reply:
I got that from [the official docs](https://doc.rust-lang.org/std/fmt/trait.Display.html#examples).

Reply:
Having said that, I do not have a strong opinion.

Reply:
well, I expect that rust will optimize it out, just I see no reason for formatting, but also not have strong opinion and if it is common approach in implementing Format trait, it is ok for me.

Reply:
I do not know how common it is, just what I saw in the docs. Perhaps we will learn something different tomorrow 😄